File: cp264/assignments/a8/avl.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "avl.h"


AVLNODE *avl_node(RECORD data)
{
  AVLNODE *np = (AVLNODE *)malloc(sizeof(AVLNODE));
  if (np)
  {
    np->data = data;
    np->height = 1;
    np->left = NULL;
    np->right = NULL;
  }
  return np;
}


int max(int a, int b) 
{
  return (a > b)? a : b;
}


int height(AVLNODE *np)
{
// your code
    int h = 0;
    if (np) h = np->height;
    return h;
}

int balance_factor(AVLNODE *np)
{
// your code
    int b = 0;
    if (np) b = height(np->left)-height(np->right);
    return b;
}


AVLNODE *rotate_left(AVLNODE *np)
{
// your code
// your implementation
  AVLNODE *child = np->right;
	AVLNODE *temp = child->left;
	// Rearrange the nodes.
	// < your code here >
	child->left = np;
	np->right = temp; 

  np->height = max(height(np->left), height(np->right))+1;
	child->height = max(height(child->left), height(child->right))+1;
  return child;
}

AVLNODE *rotate_right(AVLNODE *np)
{
// your code

  AVLNODE *child = np->left;
	AVLNODE *temp = child->right;
	// Rearrange the nodes.
	child->right = np;
	np->left = temp;
	
	np->height = max(height(np->left), height(np->right))+1;
	child->height = max(height(child->left), height(child->right))+1;
	// Return new root.
	return (child);
}

AVLNODE *extract_smallest_node(AVLNODE **rootp) {
  AVLNODE *p = *rootp, *parent = NULL;
  if (p) {
    while (p->left) {
      parent = p;
      p = p->left;
    }
    if (parent == NULL)
      *rootp = p->right;
    else
      parent->left = p->right;
    
    p->left = NULL;
    p->right = NULL;
  }
  return p;
}

void avl_insert(AVLNODE **rootp, RECORD data)
{  
  AVLNODE *np = (AVLNODE *) malloc(sizeof(AVLNODE));
  if (np == NULL) return;
  strcpy(np->data.name, data.name);
  np->data.score = data.score;
  np->height = 1;
  np->left = NULL;
  np->right = NULL;
  
  // 1. Perform the normal BST insertion
  if (*rootp == NULL) {
    *rootp = np;
    return;
  }
  
  AVLNODE *root = *rootp;
  if (strcmp(data.name, root->data.name) < 0 )
    avl_insert(&root->left, data);
  else if (strcmp(data.name, root->data.name) > 0 )
    avl_insert(&root->right, data);
  else return ;

  // 2. update height of this root node
  root->height = 1 + max(height(root->left), height(root->right));
 
  // 3. Get the balance factor of this ancestor node to check whether this node became unbalanced
  int bf = balance_factor(root);
  
  // 4. re-balance if not balanced
  if(bf > 1){
    if(balance_factor(root->left) >= 0) *rootp = rotate_right(root);
    else{
      root->left = rotate_left(root->left);
      *rootp = rotate_right(root);
    }
  } else if(bf < -1){
    if(balance_factor(root->right) <= 0) *rootp = rotate_left(root);
    else {
      root->right = rotate_right(root->right);
      *rootp = rotate_left(root);
    }
  } 
}

void avl_delete(AVLNODE **rootp, char *name)
{
  AVLNODE *root = *rootp;
  AVLNODE *np;
  if (root == NULL) return;

  if (strcmp(name, root->data.name) == 0) {
    if (root->left == NULL && root->right == NULL) {
      free(root);
      *rootp = NULL;
    } else if (root->left != NULL && root->right == NULL) {
      np = root->left;
      free(root);
      *rootp = np;
    } else if (root->left == NULL && root->right != NULL) {
      np = root->right;
      free(root);
      *rootp = np;
    } else if (root->left != NULL && root->right != NULL) {
      np = extract_smallest_node(&root->right);
      np->left = root->left;
      np->right = root->right;
      free(root);
      *rootp = np;
    }
  } else {
    if (strcmp(name, root->data.name) < 0) {
      avl_delete(&root->left, name);
    } else {
      avl_delete(&root->right, name);
    }
  }

  // If the tree had only one node then return
  if (*rootp == NULL) return;
  root = *rootp;
    
  // STEP 2: UPDATE HEIGHT OF THE CURRENT NODE
  root->height = 1+max(height(root->left), height(root->right));

  // STEP 3: GET THE BALANCE FACTOR OF THIS NODE 
  int bf = balance_factor(root);

  // STEP 4: rebalance if not balanced
  if(bf > 1){
    if(balance_factor(root->left) >= 0) *rootp = rotate_right(root);
    else{
      root->left = rotate_left(root->left);
      *rootp = rotate_right(root);
    }
  } else if(bf < -1){
    if(balance_factor(root->right) <= 0) *rootp = rotate_left(root);
    else {
      root->right = rotate_right(root->right);
      *rootp = rotate_left(root);
    }
  } 
  

}

AVLNODE *avl_search(AVLNODE *root, char *name) {
// your code
  if(root == NULL) {
    return NULL;
  }
  if(strcmp(root->data.name, name) == 0) {
    return root;
  }
  AVLNODE *node = avl_search(root->left, name);
  if(node) {
    return node;
  }

  return avl_search(root->right, name);
}


void avl_clean(AVLNODE **rootp) {
  AVLNODE *root = *rootp;
  if (root) {
    if (root->left)
      avl_clean(&root->left);
    if (root->right)
      avl_clean(&root->right);
    free(root);
  }
  *rootp = NULL;
}

File: cp213/seth6962_a04/src/cp213/AVL.java
package cp213;

/**
 * Implements an AVL (Adelson-Velsky Landis) tree. Extends BST.
 *
 * @author your name here
 * @author David Brown
 * @version 2025-09-07
 * @param <T> the data structure data type
 */
public class AVL<T extends Comparable<T>> extends BST<T> {

    /**
     * Returns the balance data of node. If greater than 1, then left heavy, if less
     * than -1, then right heavy. If in the range -1 to 1 inclusive, the node is
     * balanced. Used to determine whether to rotate a node upon insertion.
     *
     * @param node The TreeNode to analyze for balance.
     * @return A balance number.
     */
    private int balance(final TreeNode<T> node) {

	// your code here
        if (node == null) {
            return 0;
        }
        return this.nodeHeight(node.getLeft()) - this.nodeHeight(node.getRight());
    }

    /**
     * Rebalances the current node if its children are not balanced.
     *
     * @param node the node to rebalance
     * @return replacement for the rebalanced node
     */
    private TreeNode<T> rebalance(TreeNode<T> node) {

	// your code here
    	  node.updateHeight();
          int bf = balance(node);

          if (bf > 1) {
              if (balance(node.getLeft()) < 0) {
                  node.setLeft(rotateLeft(node.getLeft()));
              }
              return rotateRight(node);
          }

          if (bf < -1) {
              if (balance(node.getRight()) > 0) {
                  node.setRight(rotateRight(node.getRight()));
              }
              return rotateLeft(node);
          }

          return node;
    }

    /**
     * Performs a left rotation around node.
     *
     * @param node The subtree to rotate.
     * @return The new root of the subtree.
     */
    private TreeNode<T> rotateLeft(final TreeNode<T> node) {

	// your code here
    	TreeNode<T> newRoot = node.getRight();
        node.setRight(newRoot.getLeft());
        newRoot.setLeft(node);

        node.updateHeight();
        newRoot.updateHeight();

        return newRoot;
    }

    /**
     * Performs a right rotation around node.
     *
     * @param node The subtree to rotate.
     * @return The new root of the subtree.
     */
    private TreeNode<T> rotateRight(final TreeNode<T> node) {

	// your code here
        TreeNode<T> newRoot = node.getLeft();
        node.setLeft(newRoot.getRight());
        newRoot.setRight(node);

        node.updateHeight();
        newRoot.updateHeight();

        return newRoot;

    }

    /**
     * Auxiliary method for insert. Inserts data into this AVL. Same as BST
     * insertion with addition of rebalance of nodes.
     *
     * @param node        The current node (TreeNode).
     * @param countedItem Data to be inserted into the node.
     * @return The inserted node.
     */
    @Override
    protected TreeNode<T> insertAux(TreeNode<T> node, final CountedItem<T> countedItem) {

	// your code here
        if (node == null) {
            node = new TreeNode<T>(countedItem);
            node.getCountedItem().incrementCount();
            this.size++;
            return node;
        }

        int cmp = node.getCountedItem().compareTo(countedItem);

        if (cmp > 0) {
            node.setLeft(insertAux(node.getLeft(), countedItem));
        }
        else if (cmp < 0) {
            node.setRight(insertAux(node.getRight(), countedItem));
        }
        else {
            // duplicate
            node.getCountedItem().incrementCount();
            return node;
        }

        // Rebalance on return
        return rebalance(node);
    }

    /**
     * Auxiliary method for valid. Determines if a subtree based on node is a valid
     * subtree. An AVL must meet the BST validation conditions, and additionally be
     * balanced in all its subtrees - i.e. the difference in height between any two
     * children must be no greater than 1.
     *
     * @param node The root of the subtree to test for validity.
     * @return true if the subtree base on node is valid, false otherwise.
     */
    @Override
    protected boolean isValidAux(final TreeNode<T> node, TreeNode<T> minNode, TreeNode<T> maxNode) {

	// your code here
        if (node == null) {
            return true;
        }

        // BST property check
        if (minNode != null && node.getCountedItem().compareTo(minNode.getCountedItem()) <= 0)
            return false;

        if (maxNode != null && node.getCountedItem().compareTo(maxNode.getCountedItem()) >= 0)
            return false;

        // AVL balance property
        int bf = balance(node);
        if (bf < -1 || bf > 1)
            return false;

        // Height validity: node.height = max(left,right) + 1
        int expectedHeight = Math.max(nodeHeight(node.getLeft()), nodeHeight(node.getRight())) + 1;
        if (node.getHeight() != expectedHeight)
            return false;

        // Check subtrees
        return isValidAux(node.getLeft(), minNode, node)
                && isValidAux(node.getRight(), node, maxNode);

    }

    /**
     * Determines whether two AVLs are identical.
     *
     * @param target The AVL to compare this AVL against.
     * @return true if this AVL and target contain nodes that match in position,
     *         data, count, and height, false otherwise.
     */
    public boolean equals(final AVL<T> target) {
	return super.equals(target);
    }

    /**
     * Auxiliary method for remove. Removes data from this BST. Same as BST removal
     * with addition of rebalance of nodes.
     *
     * @param node        The current node (TreeNode).
     * @param countedItem Data removed from the tree.
     * @return The replacement node.
     */
    @Override
    protected TreeNode<T> removeAux(TreeNode<T> node, final CountedItem<T> countedItem) {

	// your code here
        if (node == null) {
            return null;
        }

        int cmp = node.getCountedItem().compareTo(countedItem);

        if (cmp > 0) {
            node.setLeft(removeAux(node.getLeft(), countedItem));
        }
        else if (cmp < 0) {
            node.setRight(removeAux(node.getRight(), countedItem));
        }
        else {
            // found node
            if (node.getCountedItem().getCount() > 1) {
                node.getCountedItem().decrementCount();
                return node;
            }

            // leaf
            if (node.getLeft() == null && node.getRight() == null) {
                this.size--;
                return null;
            }
            // one child
            else if (node.getLeft() == null) {
                this.size--;
                return node.getRight();
            }
            else if (node.getRight() == null) {
                this.size--;
                return node.getLeft();
            }
            else {
                // two children → replace node with successor node
                TreeNode<T> succ = node.getRight();
                while (succ.getLeft() != null) {
                    succ = succ.getLeft();
                }

                // Remove successor from its old location
                node.setRight(removeAux(node.getRight(), succ.getCountedItem()));

                // Attach children to successor
                succ.setLeft(node.getLeft());
                succ.setRight(node.getRight());

                node = succ;

                this.size--;
            }
        }

        // Rebalance on way up
        return rebalance(node);
    }

}
